level.py

```diff
@@ -53,15 +88,15 @@ def parse_line(self, filename, line):
             raise IOError("Invalid format in {}, could not find separator(:) on line {}".format(filename, line))
 
         try:
-            time = float(line[:separator])
+            move = int(line[:separator])
         except ValueError:
             raise IOError("Invalid format in {}, could not parse float from {}".format(filename, line[:separator]))
 
-        if time in self._events.keys():
-            raise IOError("Invalid format in {}, timestamp {} seen again on line {}".format(filename, time, line))
+        if move == self._event_get(-1)[0]:
```

Comment:
I'm really confused by this line, it feels like it should be a peek rather than -1 (which then needs other strange-looking code specifically to account for it). If we are using a list as a stack then just mutate the list (it can be a deepcopy if we still want to original around) https://docs.python.org/3.1/tutorial/datastructures.html#using-lists-as-stacks

Reply:
yeah this solution is super messy. There are 3 things I was trying to do:
1. Peek and Pop as FIFO queue not LIFO stack for progressively passing events to model.
2. LIFO Peek. The -1 is to check duplicate (and soon out of order) moves - maybe we just do this another way. Also List doesn't have a peek method anyway so you still have to do a -1 and catch errors
3. has to work on an empty list / exceeding end index 

self._current_event variable was an attempt to solve 1 in constant time, which is harder for FIFO than LIFO. And the rest of the mess was to deal 3. I can see these alternatives: 
* import deque to efficiently pop from the start 
* or parse the event list backwards so we can pop off the end
* not worry about efficiency

Or am I way overcomplicating this?

Reply:
Oh this is moot because we can make this a dict for the discreet version where move number has to match exactly

Reply:
Agreed it's moot, in general whatever behavior you want for throwing/peeking/queue-over-stack you can just use len(list) == 0, list[-1], list.pop(-1)